src/compose/MentionWarnings.js

```diff
@@ -0,0 +1,140 @@
+/* @flow strict-local */
+
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+
+import type { Auth, Stream, Dispatch, Narrow, UserOrBot, Subscription } from '../types';
+import { getActiveUsersById, getAuth } from '../selectors';
+import { isPrivateNarrow } from '../utils/narrow';
+import * as api from '../api';
+
+import AnimatedScaleComponent from '../animation/AnimatedScaleComponent';
+import MentionedUserNotSubscribed from '../message/MentionedUserNotSubscribed';
+
+type State = {|
+  unsubscribedMentions: UserOrBot[],
+|};
+
+type SelectorProps = {|
+  auth: Auth,
+  usersById: Map<number, UserOrBot>,
+|};
+
+type Props = $ReadOnly<{|
+  narrow: Narrow,
+  stream: Subscription | {| ...Stream, in_home_view: boolean |},
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class MentionWarnings extends PureComponent<Props, State> {
+  state = {
+    unsubscribedMentions: [],
+  };
+
+  /**
+   * Check whether the message text entered by the user contains
+   * an @-mention to a user unsubscribed to the current stream, and if
+   * so, shows a warning.
+   *
+   * This function is expected to be called by `ComposeBox` using a ref
+   * to this component.
+   *
+   * @param message The message text entered by the user in the compose box.
+   */
+  handleMentionSubscribedCheck = async (message: string) => {
+    const { usersById, narrow, auth, stream } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return;
+    }
+    const unformattedMessage = message.split('**')[1];
+
+    // We skip user groups, for which autocompletes are of the form
+    // `*<user_group_name>*`, and therefore, message.split('**')[1]
+    // is undefined.
+    if (unformattedMessage === undefined) {
+      return;
+    }
+    const [userFullName, userId] = unformattedMessage.split('|');
+    const unsubscribedMentions = this.state.unsubscribedMentions.slice();
+    let mentionedUser: UserOrBot;
+
+    // eslint-disable-next-line no-unused-vars
+    for (const [id, user] of usersById) {
+      if (userId !== undefined) {
+        if (user.user_id === userId) {
+          mentionedUser = user;
+          break;
+        }
+      } else if (user.full_name === userFullName) {
+        mentionedUser = user;
+        break;
+      }
+    }
+    if (!mentionedUser || unsubscribedMentions.includes(mentionedUser)) {
+      return;
```

Comment:
This has a bug because it's going to compare the user objects for identity. If `unsubscribedMentions` contains an object describing user 1234, and `mentionedUser` is user 1234, but they're different objects, then it will not see the match. One way this could happen is if the user object gets updated in Redux as a result of that user changing something (like their avatar URL) after they've already been added to `unsubscribedMentions` -- then `unsubscribedMentions` will still have the old object.

Instead, the comparison should just look at user IDs.

Probably the simplest solution is that `unsubscribedMentions` should just *be* an array of user IDs. When we need full user objects, we can get them by looking them up in `usersById` (with `usersById.get`, as mentioned in another comment.)